### PR TITLE
Add tests for Tickets filter logic; closes #2103.

### LINKED ISSUE #2103: Add tests for Tickets filter logic.

### PR DESCRIPTION
**Pull request by nicolad**
## Problem this Pull Request solves
Closes: https://github.com/eventespresso/event-espresso-core/issues/2103

## Checklist

* [ ] I have read the documentation relating to systems affected by this pull request, see https://github.com/eventespresso/event-espresso-core/tree/master/docs
* [ ] User input is adequately validated and sanitized
* [ ] all publicly displayed strings are internationalized (usually using `esc_html__()`, see https://codex.wordpress.org/I18n_for_WordPress_Developers)
* [ ] My code is tested.
* [ ] My code follows the Event Espresso code style.
* [ ] My code has proper inline documentation.
* [ ] My code accounts for when the site is in Maintenance Mode (MM2 especially disallows usage of models)

### CODE REVIEW

File: assets/prototype/domain/eventEditor/tickets/ticketList/filterBar/controls/ShowTicketsControl.tsx [hunk 1]
```diff
@@ -14,62 +19,64 @@ import { __ } from '@eventespresso/i18n';
...
-						value: 'next-on-sale-or-pending-only',
-						label: __('next on sale or sale pending only', 'event_espresso'),
+						value: ShowTickets.archivedOnly,
+						label: __('archived tickets only', 'event_espresso'),
 					},
 					{
-						value: 'sold-out-only',
-						label: __('sold out tickets only', 'event_espresso'),
+						value: ShowTickets.below50Sold,
+						label: __('tickets with less than 50% sold', 'event_espresso'),
```
Comment (manzoorwanijk) [inline assets/prototype/domain/eventEditor/tickets/ticketList/filterBar/controls/ShowTicketsControl.tsx]: Nice refactor

File: assets/prototype/domain/eventEditor/tickets/ticketList/filterBar/useTicketListFilterState/index.test.ts [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+import { DisplayDates, ShowTickets, TicketsSortedBy } from '../../../../data/ticket/types';
+
+import useTicketListFilterState from './index';
+import { tickets } from '../../../../../shared/predicates/tickets/test/data';
+
+test('useTicketListFilterState initial state result', () => {
+	const { result } = renderHook(() => useTicketListFilterState(tickets));
+	const currentResult = result.current;
+
+	expect(Object.keys(currentResult).length).toBe(10);
```
Comment (manzoorwanijk) [inline assets/prototype/domain/eventEditor/tickets/ticketList/filterBar/useTicketListFilterState/index.test.ts]: 👍

File: assets/prototype/domain/eventEditor/tickets/ticketList/filterBar/useTicketListFilterState/index.ts [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+		setTicketsSortedBy,
+		toggleIsChained,
+	};
+};
+
+interface Action {
+	type: ActionType;
+}
+
+const reducer = (state, action) => {
```
Comment (manzoorwanijk) [inline assets/prototype/domain/eventEditor/tickets/ticketList/filterBar/useTicketListFilterState/index.ts]: Any reasons for not typefying it or that is going to be a separate PR?

File: assets/prototype/application/services/entityTextFieldSearch/index.test.ts [hunk 4]
```diff
@@ -77,3 +85,19 @@ test('Search with two results', () => {
...
+	expect(searchResult).toEqual(datetimes);
+});
+
+test('Should return empty array if empty entity list is provided and no search term is provided ', () => {
+	const searchResult = entityTextFieldSearch({ entities: [], searchText: '' });
+
+	expect(Array.isArray(searchResult)).toBe(true);
+	expect(searchResult.length).toBe(0);
+	expect(searchResult).toEqual([]);
+});
```
Comment (nicolad) [inline assets/prototype/application/services/entityTextFieldSearch/index.test.ts]: Changes from this file were pulled into a related PR: https://github.com/eventespresso/event-espresso-core/pull/2157

File: assets/prototype/application/services/entityTextFieldSearch/index.ts [hunk 5]
```diff
@@ -1,3 +1,8 @@
+/**
```
Comment (nicolad) [inline assets/prototype/application/services/entityTextFieldSearch/index.ts]: Changes from this file were pulled into a related PR: https://github.com/eventespresso/event-espresso-core/pull/2157

File: assets/prototype/domain/eventEditor/tickets/ticketList/filterBar/useTicketListFilterState/index.ts (hunk 3 above)
Comment (nicolad) [reply assets/prototype/domain/eventEditor/tickets/ticketList/filterBar/useTicketListFilterState/index.ts]: This PR got too hairy.
I will come back on this area when I will couple to UI and will do some refactoring and types also will be added.

**Comment by nicolad** [comment]
@manzoorwanijk @tn3rb the initial scope of the related issue wasn't very specific and this is why I moved this to review.
There are some things left to be done which I will do when I will couple this logic to UI.

File: assets/prototype/domain/eventEditor/data/queries/tickets/useTicketItem.ts [hunk 6]
```diff
@@ -6,7 +6,8 @@ import { EntityItemProps, ReadQueryOptions } from '../types';
 
 const useTicketItem = ({ id }: EntityItemProps): Ticket => {
 	const client = useApolloClient();
-	let data: any, ticket: Ticket;
+	let data: any;
+	let ticket: Ticket;
```
Comment (nicolad) [inline assets/prototype/domain/eventEditor/data/queries/tickets/useTicketItem.ts]: This enhances readability.

File: assets/prototype/domain/shared/predicates/tickets/filters/percentSoldAtOrAbove/index.ts [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+	maxQuantity: number;
+	tickets: Ticket[];
+};
+
+/**
+ * @param {Object} ticket event ticket object
+ * @param {number} maxQuantity
+ * @return {boolean} true if sold/qty >= maxQuantity
+ */
+const percentSoldAtOrAbove = ({ maxQuantity, tickets }: PercentSoldAtOrAboveProps) => {
```
Comment (tn3rb) [inline assets/prototype/domain/shared/predicates/tickets/filters/percentSoldAtOrAbove/index.ts]: Not your fault since this was originally my code, but I'm thinking something like `percentage` might be more clear than `maxQuantity` for the variable name, since that's not really a quantity.
I got confused by the tests when I saw `maxQuantity` as an integer like 75 or 90 and was thinking "75 tickets? 90 tickets?" whereas if i had named that var `percentage`, then it could not be so easily confused.

What do you think?

File: assets/prototype/domain/shared/predicates/tickets/filters/sortTicketsList.ts [hunk 8]
```diff
@@ -1,23 +1,24 @@
...
-		case DatesSortedBy.chronologically:
-			return sortWith([ascend(prop('name')), ascend(prop('id'))], tickets);
-		case DatesSortedBy.byName:
-			return sortWith([ascend(prop('name'))], tickets);
-		case DatesSortedBy.byId:
+		case TicketsSortedBy.chronologically:
+			return sortWith([ascend(propOr(null, 'name')), ascend(prop('id'))], tickets);
+		case TicketsSortedBy.byName:
+			return sortWith([ascend(propOr(null, 'name'))], tickets);
+		case TicketsSortedBy.byId:
```
Comment (tn3rb) [inline assets/prototype/domain/shared/predicates/tickets/filters/sortTicketsList.ts]: `TicketsSortedBy.chronologically`, `TicketsSortedBy.byName`, `TicketsSortedBy.byId` all look a little weird because of the duplicated use of "by"

What about following the same thing you did for the `ShowTickets` type and rename `TicketsSortedBy` to `SortTickets` which would then read much better: `SortTickets.chronologically`, `SortTickets.byName`, `SortTickets.byId`

File: assets/prototype/domain/shared/predicates/tickets/filters/validFiniteQuantity.ts [hunk 9]
```diff
@@ -1,9 +1,14 @@
...
+import { Ticket } from '../../../../eventEditor/data/types';
+
 /**
  * @param {Object} ticket    event ticket object
  * @return {boolean} true if qty property is valid and NOT infinite
  */
-const validFiniteQuantity = (ticket) => {
-	return ticket.qty !== 'INF' && ticket.qty !== Infinity && parseInt(ticket.qty, 10) > 0;
+const validFiniteQuantity = (ticket: Ticket): boolean => {
+	return ticket.quantity !== Infinity && Math.round(ticket.quantity) > 0;
```
Comment (tn3rb) [inline assets/prototype/domain/shared/predicates/tickets/filters/validFiniteQuantity.ts]: nice

also just noticed this method while checking the data type for `Infinity`:
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Number/isFinite

File: assets/prototype/domain/shared/predicates/tickets/isPending.ts [hunk 10]
```diff
@@ -1,17 +1,18 @@
...
 /**
  * @function
  * @param {Object} ticket object
  * @return {boolean} 	true if ticket is not yet available for purchase,
  * 						but will be at some date in the future
  */
-const isPending = ({ startDate }) => {
-	return differenceInMinutes(startDate, now) > 0;
+const isPending = ({ startDate }: Ticket) => {
+	return differenceInMinutes(parseISO(startDate), now) > 0;
```
Comment (tn3rb) [inline assets/prototype/domain/shared/predicates/tickets/isPending.ts]: needs typing for return value

File: assets/prototype/domain/shared/predicates/tickets/isArchived.ts [hunk 11]
```diff
@@ -1,10 +1,16 @@
...
  * @function
- * @param {Object} ticket model object
+ * @param {Object} ticket object
  * @return {boolean} true if ticket is archived
  */
-const isArchived = (ticket) => {
-	return ticket.deleted;
+const isArchived = (ticket: Ticket) => {
+	// TODO: update according to this when it's ready: https://github.com/eventespresso/event-espresso-core/issues/2089
+	return ticket['deleted'];
```
Comment (tn3rb) [inline assets/prototype/domain/shared/predicates/tickets/isArchived.ts]: needs typing for return value

File: assets/prototype/domain/shared/predicates/tickets/filters/validInfiniteQuantity.ts [hunk 12]
```diff
@@ -1,9 +1,20 @@
...
+
 /**
  * @param {Object} ticket    event ticket object
  * @return {boolean} true if qty property is valid and unlimited
  */
-const validInfiniteQuantity = (ticket) => {
-	return ticket.qty === 'INF' || ticket.qty === Infinity;
+const validInfiniteQuantity = (ticket: Ticket) => {
+	const isNumber = is(Number, ticket.quantity); // This check has been added because qty is optional in Ticket type.
+	return isNumber && ticket.quantity === Infinity;
```
Comment (tn3rb) [inline assets/prototype/domain/shared/predicates/tickets/filters/validInfiniteQuantity.ts]: needs typing for return value

File: assets/prototype/domain/eventEditor/tickets/ticketList/filterBar/useTicketListFilterState/index.ts [hunk 13]
```diff
@@ -0,0 +1,101 @@
...
+		processedTickets: [],
+		showTickets: ShowTickets.all,
+		tickets,
+		ticketsSortedBy: TicketsSortedBy.chronologically,
+	};
+	const [state, dispatch] = useReducer(reducer, initialState);
+
+	const setDisplayTicketDate = (displayTicketDate) => {
+		dispatch({
+			type: ActionType.SET_DISPLAY_TICKET_DATE,
```
[the remark this answers is not in the document]
Comment (nicolad) [reply assets/prototype/domain/eventEditor/tickets/ticketList/filterBar/useTicketListFilterState/index.ts]: That action is being used in `DisplayTicketsControl`.

File: assets/prototype/domain/shared/predicates/tickets/filters/percentSoldAtOrAbove/index.test.ts [hunk 14]
```diff
@@ -0,0 +1,147 @@
...
+		maxQuantity: 10,
+		tickets: [
+			{
+				id: '1',
+				quantity: Infinity,
+			},
+			{
+				id: '2',
+				quantity: 0,
+			},
```
[the remark this answers is not in the document]
Comment (nicolad) [reply assets/prototype/domain/shared/predicates/tickets/filters/percentSoldAtOrAbove/index.test.ts]: TS isn't allowing me to set string values to a typed property :smile: 
image

[the remark this answers is not in the document]
Comment (nicolad) [reply assets/prototype/domain/shared/predicates/tickets/filters/percentSoldAtOrAbove/index.test.ts]: `percentSoldAtOrAbove` has typed args

File: assets/prototype/domain/shared/predicates/tickets/filters/percentSoldAtOrAbove/index.test.ts [hunk 15]
```diff
@@ -0,0 +1,147 @@
...
+
+	expect(filteredTickets.length).toBe(3);
+	expect(filteredTickets[0].id).toBe('1');
+	expect(filteredTickets[1].id).toBe('3');
+	expect(filteredTickets[2].id).toBe('4');
+});
+
+test('Should filter out tickets based on above75Sold show type', () => {
+	const filteredTickets = percentSoldAtOrAbove({
+		maxQuantity: 75,
```
[the remark this answers is not in the document]
Comment (nicolad) [reply assets/prototype/domain/shared/predicates/tickets/filters/percentSoldAtOrAbove/index.test.ts]: removed, thx

File: assets/prototype/domain/shared/predicates/tickets/filters/percentSoldAtOrAbove/index.ts (hunk 7 above)
Comment (nicolad) [reply assets/prototype/domain/shared/predicates/tickets/filters/percentSoldAtOrAbove/index.ts]: Agree, `percentage` is a more accurate description.
https://github.com/eventespresso/event-espresso-core/pull/2150/commits/2fbe27ca53d52995cb7d2c3a59d45a5a9c882027

File: assets/prototype/domain/shared/predicates/tickets/filters/sortTicketsList.ts (hunk 8 above)
Comment (nicolad) [reply assets/prototype/domain/shared/predicates/tickets/filters/sortTicketsList.ts]: Poetry level suggestion, thx :bowing_man: 
https://github.com/eventespresso/event-espresso-core/pull/2150/commits/b8bde44757816a7d04453c466b4253184d74ce7c

File: assets/prototype/domain/shared/predicates/tickets/filters/validFiniteQuantity.ts (hunk 9 above)
Comment (nicolad) [reply assets/prototype/domain/shared/predicates/tickets/filters/validFiniteQuantity.ts]: Adjusted: https://github.com/eventespresso/event-espresso-core/pull/2150/commits/20e1781c6e2165ff384a967bcd31d7883a2eae84

Comment (nicolad) [reply assets/prototype/domain/shared/predicates/tickets/filters/validFiniteQuantity.ts]: Also added some checking for sanity:
image

File: assets/prototype/domain/shared/predicates/tickets/filters/validFiniteQuantity.ts [hunk 16]
```diff
@@ -8,7 +13,11 @@ import { Ticket } from '../../../../eventEditor/data/types';
  * @return {boolean} true if qty property is valid and NOT infinite
  */
 const validFiniteQuantity = (ticket: Ticket): boolean => {
-	return ticket.quantity !== Infinity && Math.round(ticket.quantity) > 0;
+	const { quantity } = ticket;
+	const isNumber = is(Number, quantity);
```
Comment (nicolad) [inline assets/prototype/domain/shared/predicates/tickets/filters/validFiniteQuantity.ts]: Added this check because `quantity` is optional in `Ticket` interface.

File: assets/prototype/domain/shared/predicates/tickets/isPending.ts (hunk 10 above)
Comment (nicolad) [reply assets/prototype/domain/shared/predicates/tickets/isPending.ts]: Added: https://github.com/eventespresso/event-espresso-core/pull/2150/commits/b8bebcf79946133a296c6ad4cccd52bddb17ea85

File: assets/prototype/domain/shared/predicates/tickets/filters/validInfiniteQuantity.ts (hunk 12 above)
Comment (nicolad) [reply assets/prototype/domain/shared/predicates/tickets/filters/validInfiniteQuantity.ts]: Added: https://github.com/eventespresso/event-espresso-core/pull/2150/commits/b8bebcf79946133a296c6ad4cccd52bddb17ea85

File: assets/prototype/domain/shared/predicates/tickets/isArchived.ts (hunk 11 above)
Comment (nicolad) [reply assets/prototype/domain/shared/predicates/tickets/isArchived.ts]: Added: https://github.com/eventespresso/event-espresso-core/pull/2150/commits/b8bebcf79946133a296c6ad4cccd52bddb17ea85

File: assets/prototype/domain/eventEditor/tickets/ticketList/filterBar/useTicketListFilterState/index.ts (hunk 13 above)
[the remark this answers is not in the document]
Comment (tn3rb) [reply assets/prototype/domain/eventEditor/tickets/ticketList/filterBar/useTicketListFilterState/index.ts]: ya the action is not what I was referring to, there was some other commented out code there when I wrote my comment, but it is not there now!

File: assets/prototype/domain/shared/predicates/tickets/filters/percentSoldAtOrAbove/index.test.ts (hunk 14 above)
[the remark this answers is not in the document]
Comment (tn3rb) [reply assets/prototype/domain/shared/predicates/tickets/filters/percentSoldAtOrAbove/index.test.ts]: what the? this file wasn't `.ts` when I wrote that comment ?!?!? GitHub caching is getting frustrating!

File: assets/prototype/domain/eventEditor/tickets/ticketList/filterBar/useTicketListFilterState/index.ts [hunk 17]
```diff
@@ -0,0 +1,101 @@
...
+ * Internal dependencies
+ */
+import filters from '../../../../../shared/predicates/tickets/filters';
+import { DisplayDates, ShowTickets, SortTickets } from '../../../../data/ticket/types';
+import { Ticket } from '../../../../data/types';
+
+enum ActionType {
+	SET_DISPLAY_TICKET_DATE = 'SET_DISPLAY_TICKET_DATE',
+	SET_SHOW_TICKETS = 'SET_SHOW_TICKETS',
+	SET_TICKETS_SORTED_BY = 'SET_TICKETS_SORTED_BY',
```
Comment (tn3rb) [inline assets/prototype/domain/eventEditor/tickets/ticketList/filterBar/useTicketListFilterState/index.ts]: maybe change this to match the name change for the enum?

```suggestion
	SET_SORT_TICKETS = 'SET_SORT_TICKETS',
```

File: assets/prototype/domain/shared/predicates/datetimes/filters/index.ts [hunk 18]
```diff
@@ -18,7 +19,7 @@ import upcomingOnly from './upcomingOnly';
 export const now = new Date();
 
 interface FilterDates {
-	dates: any[];
+	dates: Datetime[];
```
Comment (tn3rb) [inline assets/prototype/domain/shared/predicates/datetimes/filters/index.ts]: 👍

File: assets/prototype/domain/shared/predicates/tickets/filters/percentSoldAtOrAbove/index.test.ts (hunk 14 above)
[the remark this answers is not in the document]
Comment (tn3rb) [reply assets/prototype/domain/shared/predicates/tickets/filters/percentSoldAtOrAbove/index.test.ts]: ya awesome... I'm loving TS so far

File: assets/prototype/domain/eventEditor/tickets/ticketList/filterBar/useTicketListFilterState/index.ts (hunk 17 above)
Comment (nicolad) [reply assets/prototype/domain/eventEditor/tickets/ticketList/filterBar/useTicketListFilterState/index.ts]: Done, thx.
